### PR TITLE
feat(storage): introduce streaming_table write and point-get

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

As discussed, we have decided to refactor relational layer. [New Design of Relational Table](https://singularity-data.quip.com/WyWoAAiAmjwB/RFC-New-Design-of-Relational-Table)

In the future, folder `table` will be split into two folders
- `streaming_table`, which includes `StateTable` and `MemTable`, is responsible  for the streaming executors writing/read from state store
- `batch_table`, which includes `StorageTable`, is used by batch mode.

This PR:
- creates streaming_table
-  **change `MemTable` to store Row in `vec<u8>`,** and serialize `Row` before writing into `MemTable`
- change materialize executor to use streaming table
- The new `StateTable` is no need to implement `RS: RowSerde`

Note that `Scan`(`StateTableIter`) will be implemented in later PRs to avoid a PR hard to review.
 

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
part of https://github.com/singularity-data/risingwave/issues/4557

### CODE REVIEW

**Review by fuyufjh — APPROVED** [review]
Generally LGTM. PTAL @st1page ❤️

File: src/storage/src/streaming_table/mod.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+use risingwave_common::buffer::{Bitmap, BitmapBuilder};
+use risingwave_common::catalog::Schema;
+use risingwave_common::types::VIRTUAL_NODE_COUNT;
+
+use crate::error::StorageResult;
+use crate::streaming_table::state_table::DEFAULT_VNODE;
+
+/// Represents the distribution for a specific table instance.
+#[derive(Debug)]
+pub struct Distribution {
```
Comment (hzxa21) [inline src/storage/src/streaming_table/mod.rs:35]: Will the `Distribution` and `TableIter` in this file be used by the future batch_table? If yes, it is better to put it outside of `streming_table`. How about organizing the folder like:
```
├──storage
│  ├── table
│     ├── <common stuff> ...
│     ├── streaming
│       ├── ....
│     ├── batch
│       ├── ....
```

File: src/storage/src/streaming_table/mem_table.rs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+#[derive(Clone)]
+pub enum RowOp {
+    Insert(Vec<u8>),
+    Delete(Vec<u8>),
+    Update((Vec<u8>, Vec<u8>)),
+}
+
+/// `MemTable` is a buffer for modify operations without encoding
+#[derive(Clone)]
+pub struct MemTable {
```
Comment (hzxa21) [inline src/storage/src/streaming_table/mem_table.rs:27]: See comments below. Will `MemTable` be used by batch table in the future?

File: src/storage/src/streaming_table/state_table.rs [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+    dist_key_indices: Vec<usize>,
+
+    /// Indices of distribution key for computing vnode.
+    /// Note that the index is based on the primary key columns by `pk_indices`.
+    dist_key_in_pk_indices: Vec<usize>,
+
+    /// Virtual nodes that the table is partitioned into.
+    ///
+    /// Only the rows whose vnode of the primary key is in this set will be visible to the
+    /// executor. For READ_WRITE instances, the table will also check whether the writed rows
```
Comment (hzxa21) [inline src/storage/src/streaming_table/state_table.rs:87]: Now StateTable is always READ_WRITE, please modify the comment accordingly.

File: src/storage/src/streaming_table/mod.rs (hunk 1 above)
Comment (wcy-fdu) [reply src/storage/src/streaming_table/mod.rs:35]: Make sense to me.

File: src/storage/src/streaming_table/state_table.rs (hunk 3 above)
Comment (wcy-fdu) [reply src/storage/src/streaming_table/state_table.rs:87]: Will update these comments later.

**Comment by wcy-fdu** [comment]
Will continue to refactor the encoding part.

**Comment by st1page** [comment]
Some offline discussion: the RowSerialize interface abstraction is for cell-based encoding and not suitable for our current demand, we will use a more clear and lighter method for it and add back if we really need

File: src/storage/src/streaming_table/mod.rs (hunk 1 above)
Comment (wcy-fdu) [reply src/storage/src/streaming_table/mod.rs:35]: Let's apply the new folder after finishing refactoring of `streaming_table` and `batch_table`.

File: src/storage/src/row_serde/row_serde_util.rs [hunk 4]
```diff
@@ -111,3 +113,49 @@ pub fn parse_raw_key_to_vnode_and_key(raw_key: &[u8]) -> (VirtualNode, &[u8]) {
...
+/// todo(wcy-fdu): remove `RowSerde` trait after all executors using streaming table.
+pub fn serialize(vnode: VirtualNode, pk: &[u8], row: Row) -> Result<(Vec<u8>, Vec<u8>)> {
+    let mut value_bytes = vec![];
+    let key = [vnode.to_be_bytes().as_slice(), pk].concat();
+    for cell in &row.0 {
+        value_bytes.extend(serialize_datum(cell)?);
+    }
+    let res = (key, value_bytes);
+    Ok(res)
+}
```
Comment (st1page) [inline src/storage/src/row_serde/row_serde_util.rs:127]: after read the current code in state table, I think maybe we can just change it to just serialize the value. the pk part we can use the `StateTable::serialize_pk_with_vnode`. Also after that we can remove `compute_vnode_by_row`

File: src/storage/src/streaming_table/state_table.rs [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+
+    /// Delete a row from state table. Must provide a full row of old value corresponding to the
+    /// column desc of the table.
+    pub fn delete(&mut self, old_value: Row) -> StorageResult<()> {
+        let mut datums = vec![];
+        for pk_index in self.pk_indices() {
+            datums.push(old_value.index(*pk_index).clone());
+        }
+        let pk = Row::new(datums);
+        let pk_bytes = serialize_pk(&pk, self.pk_serializer());
```
Comment (st1page) [inline src/storage/src/streaming_table/state_table.rs:366]:
```suggestion
        let mut datums = vec![];
        for pk_index in self.pk_indices() {
            datums.push(old_value.index(*pk_index).clone());
        }
        let pk = Row::new(self.pk_indices().map(|pk_index| old_value.index(*pk_index).clone()).collect());
```
Also, we might need a method `StateTable::get_pk_from_row`

**Review by st1page — APPROVED** [review]
LGTM

File: src/storage/src/streaming_table/mem_table.rs [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    pub fn update(&mut self, pk: Vec<u8>, old_value: Vec<u8>, new_value: Vec<u8>) {
+        let entry = self.buffer.entry(pk);
+        match entry {
+            Entry::Vacant(e) => {
+                e.insert(RowOp::Update((old_value, new_value)));
+            }
+            Entry::Occupied(mut e) => match e.get_mut() {
+                RowOp::Insert(original_value) => {
+                    debug_assert_eq!(original_value, &old_value);
+                    e.insert(RowOp::Update((old_value, new_value)));
```
Comment (BugenZhao) [inline src/storage/src/streaming_table/mem_table.rs:119]: Should it be
```rust
e.insert(RowOp::Insert(new_value));
```
?

Comment (hzxa21) [reply src/storage/src/streaming_table/mem_table.rs:119]: +1. Good catch.

**Pull request merged**